### PR TITLE
cpu hot plug: improve cpu hot plug validation and tests

### PR DESCRIPTION
**Pull request by vladikr**
**What this PR does / why we need it**:
This PR is a follow-up to #9757; It addresses some immediate issues.

 - It adds validation to prevent live updating of the VMI sockets during an inflight live migration.
    Before this PR, it was theoretically possible to change the VMI sockets count before the migration hand-off; That way the target pod could be created with inadequate CPU resources.

 - Fixes some validation webhook tests.
 - Fixes migration of VMs with dedicated CPUs and defined MaxSockets
   - The original PR has an issue that prevents VM with dedicated CPUs from migrating because MaxSockets was not taken into account when calculating target CPU topology.
 
A functional test has been adjusted to verify that it's not possible to migrate hot-pluggable VMs with dedicated CPUs and that it's not possible to initiate a CPU hot plug during an inflight migration.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Review by acardace — APPROVED** [review]

**Comment by enp0s3** [comment]
/approve
/hold
@xpivarc feel free to unhold

So in the case of CPU hot update with Dedicated CPUs, the original functional test masked the issue because we were hot-updating the CPU to the maximum amount, therefore we couldn't see that we didn't take maxSockets into account when rendering the target domain XML in the migration flow.

File: pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go
Comment (xpivarc) [inline pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go:721]: Isn't this still racy? While it may be true that this returns nil another component or even the user might post a migration at the same time. I am not sure I understand this change. As suggested in the past I would just make sure the code around hotplug is resilient to this situation rather than prevent it from happening.

File: pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter_test.go
Comment (xpivarc) [inline pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter_test.go:1690]: This is a great improvement.

File: pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go
Comment (vladikr) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go:721]: This change makes sure that changing the topology will NOT be allowed when a migration object for a given VMI exists.

If the migration is posted *After*, the hot plug has changed the topology - there is no problem. The target pod will be created with the altered topology and correct resources for hot plug.
The only issue right now that exists (without this PR) is when a competing migration is posted before the target pod handoff and this PR blocks it.

I'm 100% sure what resiliency are you suggesting, however, we can continue this work in a follow-up. This PR only prevents initiating the CPU hot plug when there is another inflight migration. - Same way as double migration is not allowed - hence using the same function.

Comment (xpivarc) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go:721]: All I am saying is that both the migration and the change of the spec seem to be able l go through. Certainly, the gap is being closed but not entirely.

My suggestion would be something along these lines(Warning, I did not put time into this), a vmim is being annotated with cpus that were used to calculate the target Pod. After the migration, the cpus are propagated to VMI and handler can use this in order to determine if hotplug is safe to do/or at least partial hotplug as you then don't need to reject changes.

Comment (vladikr) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go:721]: @xpivarc This needs additional thinking, from my POV. I don't see how this belongs to this PR. 
For a non-dedicated CPUs case, a lot can change in between the annotation of the VMIM and the actual plug. For the dedicated CPUs case, what we have right now is close to being enough.
I don't see a reason to hold up this PR. 
We should definitely continue the discussion, but this PR solves specific immediate issues.

**Comment by enp0s3** [comment]
/unhold

I will unhold this PR since per my understanding, complete elimination of a possible race between CPU hotplug and Live-migration is not the intention of this PR. It aims to:
* Fix the basic flow of CPU hotplug with dedicated CPU
* Fix the admitter unit tests.
* Enhance CPU hotplug e2e to validate that its not possible to do CPU hotplug while migration takes place.

Also I am not sure if the involved maintainers would be available in the next couple of days to continue the discussion.

**Comment by xpivarc** [comment]
@enp0s3 Agree with the un-hold. The second commit is important fix and the first doesn't break anything and we can re-work the part.

**Pull request merged**
